src/DurableTask.Netherite/PartitionState/OutboxState.cs

```diff
@@ -99,20 +99,28 @@ public void ConfirmDurable(Event evt)
         void Send(Batch batch)
         {
             batch.ReadyToSendTimestamp = this.Partition.CurrentTimeMs;
-            var outMessages = batch.OutgoingMessages.Count < 2 ? batch.OutgoingMessages : batch.OutgoingMessages.ToList();// prevent concurrent mod
-            batch.TotalAcksExpected = batch.OutgoingResponses.Count + outMessages.Count;
-            this.Partition.EventDetailTracer?.TraceEventProcessingDetail($"Outbox is sending {batch.OutgoingResponses.Count} responses, {outMessages.Count} messages for event id={batch.SendingEventId}");
+            batch.TotalAcksExpected = batch.OutgoingResponses.Count + batch.OutgoingMessages.Count;
+            this.Partition.EventDetailTracer?.TraceEventProcessingDetail($"Outbox is sending {batch.OutgoingResponses.Count} responses, {batch.OutgoingMessages.Count} messages for event id={batch.SendingEventId}");
 
             // now that we know the sending event is persisted, we can send the messages
-            foreach (var outresponse in batch.OutgoingResponses)
+            // we need to make a copy of the collections to safely iterate, because they are modified after the 
+            // send is confirmed, which can happen before we finish the iteration.
+
+            if (batch.OutgoingResponses.Count > 0)
             {
-                DurabilityListeners.Register(outresponse, batch);
-                this.Partition.Send(outresponse);
+                foreach (var outresponse in batch.OutgoingResponses.ToArray())
+                {
+                    DurabilityListeners.Register(outresponse, batch);
+                    this.Partition.Send(outresponse);
+                }
```

Comment:
Am I right in understanding that you're copying the responses _inline_ by calling `.ToArray()`?
If so - I think it would be better to split the copying of the collection into a separate line: otherwise, I think it's easy to miss that the `ToArray()` step is important for correctness. Please consider this:

```suggestion
                var outgoingResponses = batch.OutgoingResponses.ToArray(); //copy for safe iteration
                foreach (var outresponse in outgoingResponses )
                {
                    DurabilityListeners.Register(outresponse, batch);
                    this.Partition.Send(outresponse);
                }
```